### PR TITLE
Start using the new database wrapper. Clean up types.

### PR DESCRIPTION
**Pull request by vector-of-bool**

### CODE REVIEW

File: node/mo/utils/db.ts [hunk 1]
```diff
@@ -77,10 +66,24 @@ export async function openLinksCollection(db: Db): Promise<Collection<LinkDocume
...
     });
 }
+
+/**
+ * Convert an AsyncIterable into an array of objects
+ * @param cur An asynchronous iterable object
+ * @returns An array of all iterated elements
+ * @note Beware arbitrary large iterables!
+ */
+export async function getAll<T>(cur: AsyncIterable<T>): Promise<T[]> {
```
Comment (ShaneHarvey) [inline node/mo/utils/db.ts]: Can we use toArray() instead https://www.mongodb.com/docs/drivers/node/current/fundamentals/crud/read-operations/cursor/#return-an-array-of-all-documents

File: node/mo/utils/types.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import { NextApiHandler } from "next"
+import { ObjectId } from "mongodb";
```
Comment (ShaneHarvey) [inline node/mo/utils/types.ts]: Is ObjectId used?

File: node/mo/pages/api/links/index.ts [hunk 3]
```diff
@@ -1,23 +1,36 @@
...
-    const { MoLink } = await connect();
-    return MoLink.create(doc).then(
-      created => {
-        resp.status(202).json(created);
-      },
-      error => {
-        resp.status(400).json({ error: error.toString() });
-      },
-    );
+    if (!doc.link || !doc.alias) {
```
Comment (ShaneHarvey) [inline node/mo/pages/api/links/index.ts:16]: suggest adding:
```

import validator from "validator"

if (!validator.isURL(values.link)) {
    return resp.status(400).json({ error: 'link is not a valid URL' });
}
```

File: node/mo/pages/api/links/index.ts [hunk 4]
```diff
@@ -1,23 +1,36 @@
...
+import { MoLinkJSON, MoLinkPutJSON } from "../../../utils/types";
 
 export default createMethodHandler({
-  async get(req, resp) {
-    const { MoLink } = await connect();
-    resp.json(await MoLink.find().sort({ n: -1 }).limit(100));
+  async get(req, resp: NextApiResponse<MoLinkJSON[]>) {
+    await withLinksCollection(async coll => {
+      const links = await getAll(coll.find().sort({ n: -1 }).limit(100));
+      const json: MoLinkJSON[] = links.map(m => ({
```
Comment (ShaneHarvey) [inline node/mo/pages/api/links/index.ts]: Should we be using MDB extended JSON for encoding/decoding?

File: node/mo/pages/api/links/index.ts [hunk 5]
```diff
@@ -1,23 +1,36 @@
...
+import { getAll, openLinksCollection, withClient, withLinksCollection } from "../../../utils/db";
 import { createMethodHandler } from "../../../utils/handle"
+import { MoLinkJSON, MoLinkPutJSON } from "../../../utils/types";
 
 export default createMethodHandler({
-  async get(req, resp) {
-    const { MoLink } = await connect();
-    resp.json(await MoLink.find().sort({ n: -1 }).limit(100));
+  async get(req, resp: NextApiResponse<MoLinkJSON[]>) {
+    await withLinksCollection(async coll => {
```
Comment (ShaneHarvey) [inline node/mo/pages/api/links/index.ts]: Opening and closing a new client for each request is a big regression. Let's make a single global client like in this example: https://github.com/mongodb-developer/nextjs-with-mongodb/blob/main/lib/mongodb.js

File: node/mo/utils/db.ts (hunk 1 above)
Comment (vector-of-bool) [reply node/mo/utils/db.ts]: Oh, that's easier.

File: node/mo/pages/api/links/index.ts (hunk 4 above)
Comment (vector-of-bool) [reply node/mo/pages/api/links/index.ts]: That may work, and will get around encoding issues.

**Review by ShaneHarvey — APPROVED** [review]

File: node/mo/package.json [hunk 6]
```diff
@@ -3,16 +3,17 @@
   "version": "0.1.0",
   "private": true,
   "scripts": {
-    "dev": "next dev -p 80",
```
Comment (ShaneHarvey) [inline node/mo/package.json]: Why change this back to 3000?

Comment (vector-of-bool) [reply node/mo/package.json]: Oh, that was just for me locally. I didn't mean to add that change.

**Pull request merged**